src/mcts/search.cc

```diff
@@ -478,32 +482,67 @@ std::vector<EdgeAndNode> Search::GetBestChildrenNoTemperature(Node* parent,
-  using El = std::tuple<bool, uint64_t, float, float, EdgeAndNode>;
-  std::vector<El> edges;
+  std::vector<EdgeAndNode> edges;
   for (auto edge : parent->Edges()) {
     if (parent == root_node_ && !root_limit.empty() &&
         std::find(root_limit.begin(), root_limit.end(), edge.GetMove()) ==
             root_limit.end()) {
       continue;
     }
-    const auto WL = edge.GetWL();
-    edges.emplace_back(edge.IsTerminal() && WL > 0.0f, edge.GetN(), WL,
-                       edge.GetP(), edge);
+    edges.push_back(edge);
   }
   const auto middle = (static_cast<int>(edges.size()) > count)
                           ? edges.begin() + count
                           : edges.end();
-  std::partial_sort(edges.begin(), middle, edges.end(), std::greater<El>());
-
-  std::vector<EdgeAndNode> res;
-  std::transform(edges.begin(), middle, std::back_inserter(res),
-                 [](const El& x) { return std::get<4>(x); });
-  return res;
+  std::partial_sort(edges.begin(), middle, edges.end(),
+                    // The function returns "true" when a is preferred to b.
+                    [](const auto& a, const auto& b) {
+                      // Handle at both win/loss terminal.
+                      if (a.IsTerminal() && a.GetWL() != 0.0f &&
+                          b.IsTerminal() && b.GetWL() != 0.0f) {
+                        // Compare two terminal wins.
+                        if (a.GetWL() > 0.0f && b.GetWL() > 0.0f) {
+                          // Prefer (shorter) non-TB win over TB win.
+                          if (a.IsTbTerminal() != b.IsTbTerminal())
+                            return a.IsTbTerminal() < b.IsTbTerminal();
+                          // Prefer shorter wins.
+                          else if (a.GetM(0.0f) != b.GetM(0.0f))
+                            return a.GetM(0.0f) < b.GetM(0.0f);
+                          // Compare two terminal losses.
+                        } else if (a.GetWL() < 0.0f && b.GetWL() < 0.0f) {
+                          // Prefer (longer) TB loss over non-TB loss.
+                          if (a.IsTbTerminal() != b.IsTbTerminal())
+                            return a.IsTbTerminal() > b.IsTbTerminal();
+                          // Prefer longer losses.
+                          else if (a.GetM(0.0f) != b.GetM(0.0f))
+                            return a.GetM(0.0f) > b.GetM(0.0f);
+                          // Prefer the win / not-loss terminal.
+                        } else
+                          return a.GetWL() > 0.0f || b.GetWL() < 0.0f;
+                        // Prefer terminal win and avoid loss.
+                      } else if (a.IsTerminal() && a.GetWL() != 0.0f)
+                        return a.GetWL() > 0.0f;
+                      // Prefer other terminal loss and avoid win.
+                      else if (b.IsTerminal() && b.GetWL() != 0.0f)
+                        return b.GetWL() < 0.0f;
```

Comment:
Could these be combined somehow?

Reply:
My try from today:
```cpp
std::partial_sort(edges.begin(), middle, edges.end(),
                  [](const auto& a, const auto& b) {
                    // The function returns "true" when a is preferred to b.

                    // Lists edge types from less desirable to more desirable.
                    enum EdgeRank {
                      kTerminalLoss,
                      kTablebaseLoss,
                      kNonTerminal,  // Non terminal or terminal draw.
                      kTablebaseWin,
                      kTerminalWin,
                    };

                    auto GetEdgeRank = [](const EdgeAndNode& edge) {
                      const auto wl = edge.GetWL();
                      if (!edge.IsTerminal() || wl) return kNonTerminal;
                      if (edge.IsTbTerminal()) {
                        return wl < 0.0 ? kTablebaseLoss : kTablebaseWin;
                      }
                      return wl < 0.0 ? kTerminalLoss : kTerminalWin;
                    };

                    // If moves have different outcomes, prefer better outcome.
                    auto a_rank = GetEdgeRank(a);
                    auto b_rank = GetEdgeRank(b);
                    if (a_rank != b_rank) return a > b;

                    // Neither is terminal, use standard rule.
                    if (a_rank == kNonTerminal) {
                      // Prefer largest playouts then eval then prior.
                      if (a.GetN() != b.GetN()) return a.GetN() > b.GetN();
                      if (a.GetWL() != b.GetWL()) return a.GetWL() > b.GetWL();
                      return a.GetP() > b.GetP();
                    }

                    // Both variants are winning, prefer shortest win.
                    if (a_rank > kNonTerminal) {
                      return a.GetM(0.0f) < b.GetM(0.0f);
                    }

                    // Both variants are losing, prefer longest losses.
                    return a.GetM(0.0f) > b.GetM(0.0f);
                  });
```

Initially it also had special rules for draws, but it's hard to make it not to violate a total strict order.

E.g. suppose we have:
a. Terminal draw.
b. Non-terminal loss with N=1000.
c. Non-terminal win with N=100.

`a > b`, `c > a`, `b > c` (because `N`s are compared).
So I think it makes sense to treat as ordinary nodes.
Although maybe only order them specially when both `a` and `b` are terminals..

Reply:
With draw support:
```cpp
std::partial_sort(edges.begin(), middle, edges.end(),
                  [](const auto& a, const auto& b) {
                    // The function returns "true" when a is preferred to b.

                    // Lists edge types from less desirable to more desirable.
                    enum EdgeRank {
                      kTerminalLoss,
                      kTablebaseLoss,
                      kNonTerminal,  // Non terminal or terminal draw.
                      kTablebaseWin,
                      kTerminalWin,
                    };

                    auto GetEdgeRank = [](const EdgeAndNode& edge) {
                      const auto wl = edge.GetWL();
                      if (!edge.IsTerminal() || wl) return kNonTerminal;
                      if (edge.IsTbTerminal()) {
                        return wl < 0.0 ? kTablebaseLoss : kTablebaseWin;
                      }
                      return wl < 0.0 ? kTerminalLoss : kTerminalWin;
                    };

                    // If moves have different outcomes, prefer better outcome.
                    auto a_rank = GetEdgeRank(a);
                    auto b_rank = GetEdgeRank(b);
                    if (a_rank != b_rank) return a > b;

                    // If both are terminal draws, try to make it shorter.
                    if (a_rank == kNonTerminal && a.IsTerminal() &&
                        b.IsTerminal()) {
                      if (a.IsTbTerminal() != b.IsTbTerminal()) {
                        // Prefer non-tablebase draws.
                        return a.IsTbTerminal() < b.IsTbTerminal();
                      }
                      // Prefer shorter draws.
                      return a.GetM(0.0f) < b.GetM(0.0f);
                    }

                    // Neither is terminal, use standard rule.
                    if (a_rank == kNonTerminal) {
                      // Prefer largest playouts then eval then prior.
                      if (a.GetN() != b.GetN()) return a.GetN() > b.GetN();
                      if (a.GetWL() != b.GetWL()) return a.GetWL() > b.GetWL();
                      return a.GetP() > b.GetP();
                    }

                    // Both variants are winning, prefer shortest win.
                    if (a_rank > kNonTerminal) {
                      return a.GetM(0.0f) < b.GetM(0.0f);
                    }

                    // Both variants are losing, prefer longest losses.
                    return a.GetM(0.0f) > b.GetM(0.0f);
                  });
```

Reply:
Applied your code and fixed up some logic in a separate commit.